### PR TITLE
Add support for running namespaces in parallel

### PR DESCRIPTION
**Pull request by ascho**
Added capability to run test in parallel per namespace.

I had to use pmap construct due to lot of thread local environment setup through binding in function test-ns.
Using pooled threads breaks configured environment. by the way tests shows this error correctly!

### CODE REVIEW

**Comment by ascho** [comment]
anybody?

**Comment by weavejester** [comment]
Sometimes it takes me a day or two to get around to looking at a PR (I get a lot of them!). Also, this looks like a duplicate of #28.

**Comment by ascho** [comment]
got it, i'm sorry for insistence
yes, it is duplicate of [https://github.com/weavejester/eftest/pull/28](url)
the only difference is using pmap for parallelization to adapt to thread local setuped env (test-report and so on)

**Comment by weavejester** [comment]
Why are you using `pmap` instead of `run-in-parallel`?

**Comment by ascho** [comment]
well, this was my mistake to use pmap
for consistency i guess i should use run-in-parallel
now i am trying to do it

**Comment by ascho** [comment]
running tests for different namespaces on pooled threads

File: eftest/src/eftest/runner.clj [hunk 1]
```diff
@@ -40,14 +41,16 @@
...
-           (map #(.submit executor ^Runnable %))
-           doall
+      (->>
+          (for [f fs
+                :let [^Callable cb (cast Callable (bound-fn [] (f)))]]
+                (.submit executor cb))
+           (doall)
            (map #(.get %))
-           dorun)
+           (doall))
```
Comment (weavejester) [inline eftest/src/eftest/runner.clj]: I think if we're going to change this from `dorun` to `doall`, the function name shouldn't be called `run-in-parallel`. Instead, maybe we want something more like:

```clojure
(defn- ^Callable bound-callback [f]
  (cast Callable (bound-fn* f)))

(defn- ^ExecutorService threadpool-executor []
  (Executors/newFixedThreadPool (+ 2 (.availableProcessors (Runtime/getRuntime))))

(defn- pcalls* [fs]
  (let [executor (threadpool-executor)]
    (->> fs
         (map #(.submit executor (bound-callback %)))
         (doall)
         (map #(.get %)))))

(defn- pmap* [f xs]
  (pcalls* (map (fn [x] #(f x)) xs)))

(defn- test-all [vars opts]
  (let [mapf (if (:multithread-ns? opts true) pmap* map)]
    (->> (group-by (comp :ns meta) vars)
         (mapf (fn [[ns vars]] (test-ns ns vars opts)))
         (apply merge-with +)))
```

Then we replace `run-in-parallel` with `pmap*` elsewhere as well.

Comment (ascho) [reply eftest/src/eftest/runner.clj]: agree

**Comment by weavejester** [comment]
Thanks for the changes. I realized I forgot to add in the `try` and `finally` that shuts down the executor. Can you add that back in, otherwise we're going to end up with threads that aren't cleaned up. I think we'll also need to add in a `doall` at the end so we're not bitten by laziness.

**Comment by ascho** [comment]
done

File: eftest/src/eftest/runner.clj [hunk 2]
```diff
@@ -39,18 +40,26 @@
...
+
+(defn- pcalls* [fs]
+  (let [executor (threadpool-executor)]
     (try
       (->> fs
-           (map #(.submit executor ^Runnable %))
-           doall
-           (map #(.get %))
-           dorun)
+         (map #(.submit executor (bound-callback %)))
```
Comment (weavejester) [inline eftest/src/eftest/runner.clj]: Looks like the indentation is off here. Can you tidy it up?

**Comment by weavejester** [comment]
This looks fine. The indentation is off on a few lines, and I still need to test this, but aside from that I think we're ready to squash and merge. Please also make sure your commit message adheres to the [seven rules](https://chris.beams.io/posts/git-commit/#seven-rules).

File: eftest/src/eftest/runner.clj [hunk 3]
```diff
@@ -72,7 +81,7 @@
         (if (:multithread? opts true)
           (let [test (bound-fn* test-var)]
             (dorun (->> vars (filter synchronized?) (map test)))
-            (dorun (->> vars (remove synchronized?) (map (fn [v] #(test v))) run-in-parallel)))
+            (dorun (->> vars (remove synchronized?) (map (fn [v] #(test v))) pcalls*)))
```
Comment (weavejester) [inline eftest/src/eftest/runner.clj]: This line could be changed to:

```clojure
(dorun (->> vars (remove synchronized?) (pmap* test)))
```

I think?

File: eftest/src/eftest/runner.clj (hunk 2 above)
Comment (ascho) [reply eftest/src/eftest/runner.clj]: sure, will do it

File: eftest/src/eftest/runner.clj (hunk 3 above)
Comment (ascho) [reply eftest/src/eftest/runner.clj]: agree!

**Comment by ascho** [comment]
@weavejester  thanks for your attention!

**Comment by ascho** [comment]
done

**Comment by ascho** [comment]
sure

**Comment by weavejester** [comment]
Thanks. Ping me when you've done it.

**Comment by ascho** [comment]
done

**Comment by ascho** [comment]
forgot first letter upper case, done

**Comment by ascho** [comment]
@weavejester  done

**Comment by weavejester** [comment]
Thanks! Could you change the commit message to:

    Add :multithread-ns? option to run-tests

It'll help in future if people are searching the commits for when an option is introduced. Once that's done, I'll merge it in and cut a release. Thanks for all your work!

File: eftest/src/eftest/runner.clj [hunk 4]
```diff
@@ -69,10 +78,9 @@
                           (wrap-test-with-timer test-warn-time))]
     (once-fixtures
       (fn []
-        (if (:multithread? opts true)
-          (let [test (bound-fn* test-var)]
-            (dorun (->> vars (filter synchronized?) (map test)))
-            (dorun (->> vars (remove synchronized?) (map (fn [v] #(test v))) run-in-parallel)))
+        (when (:multithread? opts true)
```
Comment (weavejester) [inline eftest/src/eftest/runner.clj]: You've changed the `if` to a `when`, so instead of executing one branch or the other, it executes *both* branches.

**Comment by weavejester** [comment]
Found the problem - you've changed an `if` to a `when`, perhaps because you didn't notice that it had two clauses.

File: eftest/src/eftest/runner.clj (hunk 4 above)
Comment (ascho) [reply eftest/src/eftest/runner.clj]: my fault :( 
i didn't notice, i made change after midnight.
Thanks for review, will fix tomorrow

**Comment by ascho** [comment]
@weavejester fixed

**Comment by weavejester** [comment]
I've carried out some further testing on this. It looks like output capture has been broken by the changes. I wrote a couple of fake tests in a new namespace:

```clojure
(deftest foo
  (future (println "foo__"))
  (.start (Thread. #(println "foo_")))
  (Thread/sleep 3000)
  (binding [*out* *err*] (println "foo"))
  (is (= 1 1)))

(deftest bar
  (future (.println System/out "bar__"))
  (.start (Thread. #(.println System/out "bar_")))
  (Thread/sleep 2000)
  (.println System/err "bar")
  (is (= 1 1)))
```

The `(.println System/err "bar")` line isn't being captured correctly.

**Comment by ascho** [comment]
i will have a look into this issue
thanks for tests!

**Comment by ascho** [comment]
@weavejester i tried to reproduce issue and got this
[https://www.screencast.com/t/Ci7VbQb7q](url)

Details: I created sample project, then i added test file with your tests and run eftest from repl (screen shows eftest parameters)

I got the same behavior on the current version of eftest.

Can you provide more details about issue?

**Comment by weavejester** [comment]
Are you certain that you didn't have old dependencies or class files that would have affected the outcome? On my machine it works correctly 100% of the time on master, and 0% of the time on your branch.

However, it may be that the order of files being loaded matters, since the root cause of this problem is effectively a race condition. In my project directory, I put the file in: `eftest/test/eftest/bad_test.clj`.

As for the cause, take a look at the [with-capture](https://github.com/ascho/eftest/blob/10f51cbd55a3cf7efeaf3f2e536bfb025bc48fe5/eftest/src/eftest/output_capture.clj#L51) macro; it uses `with-redefs`, which isn't limited to a single thread. Now take a look at [where it's used](https://github.com/ascho/eftest/blob/10f51cbd55a3cf7efeaf3f2e536bfb025bc48fe5/eftest/src/eftest/runner.clj#L91). It's used in `test-ns`, which will work fine if `test-ns` is executed in serial, but will fail (sometimes) if it's executed in parallel. You want to move the `with-capture` up one level, I think, to `test-all`.

**Comment by ascho** [comment]
@weavejester , thanks for advice!
I reproduced issue successfully.
My changes are following:

```
(defn- test-all [vars {:as opts :keys [capture-output?] :or {capture-output? true}}]
  (let [mapf (if (:multithread-ns? opts true) pmap* map)
        f (fn []
            (->> (group-by (comp :ns meta) vars)
                 (mapf (fn [[ns vars]] (test-ns ns vars opts)))
                 (apply merge-with +)))]
        (if capture-output?
          (capture/with-capture (f))
          (f))))
```
P.s. I want to clarify several moments:

1.  I didn't understand why in eftest.output-capture/with-capture uses both binding and with-redefs for *out*,*err*. Why do we need var root rebinding which has cross-thread visibility?
2. I didn't get eftest.output-capture/active-buffers var which accumulates currently used *test-buffers* :
```
(defn- doto-capture-buffer [f]
  (if *test-buffer*
    (f *test-buffer*)
    (doseq [buffer @active-buffers]
      (f buffer))))
```
In doto-capture-buffer we are calling f if *test-buffer* is not nil otherwise we are calling all buffers in active-buffers. I didn't get the situation when *test-buffer* is nil and there are some active-buffers.

**Comment by ascho** [comment]
@weavejester just to remind, my branch contains fix for capture-output issue.
when you have time please have a look!

**Comment by weavejester** [comment]
Sorry, I'm currently on vacation so my time is limited.

> I didn't understand why in eftest.output-capture/with-capture uses both binding and with-redefs for `*out*, `*err*`. Why do we need var root rebinding which has cross-thread visibility?

Because we might be testing things that use multiple threads, and some of those threads may produce output.

> I didn't get eftest.output-capture/active-buffers var which accumulates currently used `*test-buffers*`

My assumption is that this is a fallback for when `*test-buffer*` is not set. I'll take a closer look after my vacation.

File: eftest/src/eftest/runner.clj [hunk 5]
```diff
@@ -70,25 +79,27 @@
...
       (test/do-report {:type :end-test-ns, :ns ns})
       @test/*report-counters*)))
 
-(defn- test-all [vars opts]
-  (->> (group-by (comp :ns meta) vars)
-       (map (fn [[ns vars]] (test-ns ns vars opts)))
-       (apply merge-with +)))
+(defn- test-all [vars {:as opts :keys [capture-output?] :or {capture-output? true}}]
+  (let [mapf (if (:multithread-ns? opts true) pmap* map)
+        f (fn []
```
Comment (weavejester) [inline eftest/src/eftest/runner.clj]: Can you align the `let` clauses? It makes it clearer to read.

Comment (ascho) [reply eftest/src/eftest/runner.clj]: sure

**Comment by ascho** [comment]
@weavejester  thanks for feedback! sorry for your vacation interruption (

File: eftest/src/eftest/runner.clj (hunk 5 above)
Comment (ascho) [reply eftest/src/eftest/runner.clj]: fixed

**Pull request merged**